defaulting.go

```diff
@@ -0,0 +1,29 @@
+package ruleengine
+
+import "context"
+
+type DefaultingRule struct {
+	Conditions []func() bool
+	Defaulting func()
+}
+
+func ExecuteDefaulting(ctx context.Context, rules []DefaultingRule) {
+	for _, rule := range rules {
+		if rule.Conditions != nil {
+			skip := false
+
+			for _, condition := range rule.Conditions {
+				if !condition() {
+					skip = true
+					break
+				}
+			}
+
+			if skip {
+				continue
+			}
+		}
+
+		rule.Defaulting()
+	}
```

Comment:
This smells for me. Why do we need to abstract so simple flow? I understand there is some code using that already. Can you link it?

Reply:
What do you mean with smell? I think it makes sense to abstract this to not write it all over again. The same simple pieces of code are abstracted with `microerror` and `micrologger` for instance. And it makes total sense. The library provides a predefined mechanism to guide you through the complexity of the topic at hand and makes our code better in general. Because you can simply use the library. This adds to our high standards. Here are a couple of examples. Note that they are all in the AS feature branch still. 

* https://github.com/giantswarm/api/tree/feature-auto-scaling/service/cluster/creator/defaulting
* https://github.com/giantswarm/api/tree/feature-auto-scaling/service/cluster/creator/validation
* https://github.com/giantswarm/api/tree/feature-auto-scaling/service/cluster/updater/defaulting
* https://github.com/giantswarm/api/tree/feature-auto-scaling/service/cluster/updater/validation

Reply:
I don't understand why this code:

```go
	rules := []Rule{
		{
			Conditions: []func() bool{
				func() bool { return s.request.AvailabilityZones == nil },
			},
			Defaulting: func() {
				s.request.AvailabilityZones = toIntPtr(s.defaultAvailabilityZonesCount)
			},
		},
	}

	ExecuteRules(ctx, rules)

	return s.request
```

is considered better than:

```go
	if s.request.AvailabilityZones == nil {
		s.request.AvailabilityZones = toIntPtr(s.defaultAvailabilityZonesCount)
	}

	return s.request
```

It looks to me that instead of using very basic primitives we try to create some poor man's functional abstractions. I don't like that.

Reply:
Tim already explained, but I think there could be little more explanation for one who hasn't experienced our last few weeks:

@kopiczko The high level use case for this is improving manageability and maintainability of different scenarios when setting default values to requests coming to our `api`. I think the root cause of this problem is how our external API was structured and evolved and how that lead to different use cases that later proved to be problematic in terms of finding the right combination of logic rules and default values in order to be able to later process the request in our services. Now that we have been implementing couple notable features (configurable docker volume, multi-AZ, auto-scaling..) it has become pretty clear that we need to do something radical here and we now know the mid-term solution will be more or less some form of Cluster API solution in Kubernetes and getting rid of those HTTP services at least in the form we know them today. To be able to continue work on new features short-term and not lose weeks in debugging hairy if-else conditions that are spread all over the code bases (i.e. service packages within project and over multiple projects) I think this design helps us to create cohesive unit testable defaulting rules.

Reply:
Let's take more complex example (is there any more complex?). I still don't understand what is better about using this abstraction.

```go
	rules := []Rule{
		{
			Conditions: []func() bool{
				func() bool { return s.provider == "aws" },
				func() bool { return s.autoScalingEnabled },
				func() bool { return s.request.Cluster.Patch.Scaling.Min == nil },
				func() bool { return s.request.Cluster.Patch.Scaling.Max != nil },
			},
			Defaulting: func() {
				s.request.Cluster.Patch.Scaling.Min = toIntPtr(s.currentScalingMin)
			},
		},
		{
			Conditions: []func() bool{
				func() bool { return s.provider == "aws" },
				func() bool { return s.autoScalingEnabled },
				func() bool { return s.request.Cluster.Patch.Scaling.Max == nil },
				func() bool { return s.request.Cluster.Patch.Scaling.Min != nil },
			},
			Defaulting: func() {
				s.request.Cluster.Patch.Scaling.Max = toIntPtr(s.currentScalingMax)
			},
		},
		// When scaling is not set, we set it to the number of workers, if any. This
		// applies to all providers and all versions, because we run the auto scaler
		// for all new clusters and we want to track the scaling settings already
		// for clusters being created with older versions.
		{
			Conditions: []func() bool{
				func() bool { return s.request.Cluster.Patch.Scaling.IsEmpty() },
				func() bool { return len(s.request.Cluster.Patch.Workers) > 0 },
			},
			Defaulting: func() {
				s.request.Cluster.Patch.Scaling = request.Scaling{
					Min: toIntPtr(len(s.request.Cluster.Patch.Workers)),
					Max: toIntPtr(len(s.request.Cluster.Patch.Workers)),
				}
			},
		},
		{
			Conditions: []func() bool{
				func() bool { return s.request.Cluster.Patch.Scaling.IsEmpty() },
				func() bool { return len(s.request.Cluster.Patch.Workers) == 0 },
			},
			Defaulting: func() {
				s.request.Cluster.Patch.Scaling = request.Scaling{
					Min: toIntPtr(0),
					Max: toIntPtr(0),
				}
			},
		},
	}

	ExecuteRules(ctx, rules)
```

This can be replaced with following using basic control flow primitives known to any Go programmer w/o looking into internals of `ExecuteRules`.

```go
	{
		cond := true
		cond = cond && s.provider == "aws"
		cond = cond && s.autoScalingEnabled
		cond = cond && s.request.Cluster.Patch.Scaling.Min == nil
		cond = cond && s.request.Cluster.Patch.Scaling.Max != nil
		if cond {
			s.request.Cluster.Patch.Scaling.Min = toIntPtr(s.currentScalingMin)
		}
	}
	{
		cond := true
		cond = cond && s.provider == "aws"
		cond = cond && s.autoScalingEnabled
		cond = cond && s.request.Cluster.Patch.Scaling.Max == nil
		cond = cond && s.request.Cluster.Patch.Scaling.Min != nil
		if cond {
			s.request.Cluster.Patch.Scaling.Max = toIntPtr(s.currentScalingMax)
		}
	}
	// When scaling is not set, we set it to the number of workers, if any. This
	// applies to all providers and all versions, because we run the auto scaler
	// for all new clusters and we want to track the scaling settings already
	// for clusters being created with older versions.
	{
		cond := true
		cond = cond && s.request.Cluster.Patch.Scaling.IsEmpty()
		cond = cond && len(s.request.Cluster.Patch.Workers) > 0
		if cond {
			s.request.Cluster.Patch.Scaling = request.Scaling{
				Min: toIntPtr(len(s.request.Cluster.Patch.Workers)),
				Max: toIntPtr(len(s.request.Cluster.Patch.Workers)),
			}
		}
	}
	{
		cond := true
		cond = cond && s.request.Cluster.Patch.Scaling.IsEmpty()
		cond = cond && len(s.request.Cluster.Patch.Workers) == 0
		if cond {
			s.request.Cluster.Patch.Scaling = request.Scaling{
				Min: toIntPtr(0),
				Max: toIntPtr(0),
			}
		}
	}
```

> Please also consider the more complex conditions and their multitudes when arguing about what we try to achieve here.

Complex conditions make me even more worried. This abstraction checks only if all conditions are fulled. What if we need to check if any is fulled? This is just a case of replacing `&&` with `||`, but with the abstraction we may have another property like `Operation: ruleengine.OR` which would bring next level of indirection. What if relations between conditions are even more complex? This abstraction here covers only very simple case which can be solved as proposed in the code snippet above.

As I said before I don't like replacing simple control flow statements with abstractions. This introduces indirection for questionable gain. This is what I don't like the most. The examples I saw so far don't prove to me this is useful. Maintainability, extensibility and testability are nice sounding words but I don't see how any of those are improved with this abstraction. If this is about future more complex validations let's fist have the problem and then create abstraction that solves the problem instead of creating the solution upfront.

Reply:
When there is a reason to default when any condition is true, you can omit the conditions completely and only write the defaulting function. The tests show how this works. And then, as I said already, you can write your code how you prefer and nobody forces you to use this library at all. It would be cool to just unblock here because I do not see why we still talk about this. It is really not a big deal. I just want to make sure that the code base improves and this here is a tool to do so. I will merge this when there is no veto, and I do not see why there should be any.

Reply:
> When there is a reason to default when any condition is true, you can omit the conditions completely and only write the defaulting function

So we already found a case for creating workarounds.

> The tests show how this works.

Why do you want to force me to read tests when I already know the boolean operators?

I don't see it being useful or improving any of the values mentioned. I think you need a convention and not the abstraction. You have admin rights.

Reply:
I do not think we talk about the same things here and it is getting pretty silly now. When there is anything bugging you reach out to me and we can hang. I go with this now.